In [1]:
#imports
import pandas as pd
import numpy as np

#csv -> dataframe
species_df = pd.read_csv('species.csv', dtype = str)
parks_df = pd.read_csv('parks.csv')

species_df.head()
#Note: not sure why there is a 14th data column showing-there are only 13 columns of data-this is confirmed
#in the documentation for the dataset. Will come back and drop that.

Species ID             Park Name Category         Order    Family  \
0  ACAD-1000  Acadia National Park   Mammal  Artiodactyla  Cervidae   
1  ACAD-1001  Acadia National Park   Mammal  Artiodactyla  Cervidae   
2  ACAD-1002  Acadia National Park   Mammal     Carnivora   Canidae   
3  ACAD-1003  Acadia National Park   Mammal     Carnivora   Canidae   
4  ACAD-1004  Acadia National Park   Mammal     Carnivora   Canidae   

          Scientific Name                                       Common Names  \
0             Alces alces                                              Moose   
1  Odocoileus virginianus  Northern White-Tailed Deer, Virginia Deer, Whi...   
2           Canis latrans                             Coyote, Eastern Coyote   
3             Canis lupus        Eastern Timber Wolf, Gray Wolf, Timber Wolf   
4           Vulpes vulpes  Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...   

  Record Status     Occurrence  Nativeness Abundance Seasonality  \
0      Approved        Present      Native      Rare    Resident   
1      Approved        Present      Native  Abundant         NaN   
2      Approved        Present  Not Native    Common         NaN   
3      Approved  Not Confirmed      Native       NaN         NaN   
4      Approved        Present     Unknown    Common     Breeder   

  Conservation Status Unnamed: 13  
0                 NaN         NaN  
1                 NaN         NaN  
2  Species of Concern         NaN  
3          Endangered         NaN  
4                 NaN         NaN

In [2]:
#drop that extra column in species_df
species_df = species_df.iloc[: , :-1]
species_df

Species ID             Park Name        Category          Order  \
0       ACAD-1000  Acadia National Park          Mammal   Artiodactyla   
1       ACAD-1001  Acadia National Park          Mammal   Artiodactyla   
2       ACAD-1002  Acadia National Park          Mammal      Carnivora   
3       ACAD-1003  Acadia National Park          Mammal      Carnivora   
4       ACAD-1004  Acadia National Park          Mammal      Carnivora   
...           ...                   ...             ...            ...   
119243  ZION-2791    Zion National Park  Vascular Plant      Solanales   
119244  ZION-2792    Zion National Park  Vascular Plant        Vitales   
119245  ZION-2793    Zion National Park  Vascular Plant        Vitales   
119246  ZION-2794    Zion National Park  Vascular Plant  Zygophyllales   
119247  ZION-2795    Zion National Park  Vascular Plant  Zygophyllales   

                Family         Scientific Name  \
0             Cervidae             Alces alces   
1             Cervidae  Odocoileus virginianus   
2              Canidae           Canis latrans   
3              Canidae             Canis lupus   
4              Canidae           Vulpes vulpes   
...                ...                     ...   
119243      Solanaceae       Solanum triflorum   
119244        Vitaceae         Vitis arizonica   
119245        Vitaceae          Vitis vinifera   
119246  Zygophyllaceae       Larrea tridentata   
119247  Zygophyllaceae     Tribulus terrestris   

                                             Common Names Record Status  \
0                                                   Moose      Approved   
1       Northern White-Tailed Deer, Virginia Deer, Whi...      Approved   
2                                  Coyote, Eastern Coyote      Approved   
3             Eastern Timber Wolf, Gray Wolf, Timber Wolf      Approved   
4       Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...      Approved   
...                                                   ...           ...   
119243                                Cut-Leaf Nightshade      Approved   
119244                                       Canyon Grape      Approved   
119245                                         Wine Grape      Approved   
119246                                      Creosote Bush      Approved   
119247                                      Puncture Vine      Approved   

           Occurrence  Nativeness Abundance Seasonality Conservation Status  
0             Present      Native      Rare    Resident                 NaN  
1             Present      Native  Abundant         NaN                 NaN  
2             Present  Not Native    Common         NaN  Species of Concern  
3       Not Confirmed      Native       NaN         NaN          Endangered  
4             Present     Unknown    Common     Breeder                 NaN  
...               ...         ...       ...         ...                 ...  
119243        Present      Native  Uncommon         NaN                 NaN  
119244        Present      Native  Uncommon         NaN                 NaN  
119245        Present  Not Native  Uncommon         NaN                 NaN  
119246        Present      Native      Rare         NaN                 NaN  
119247        Present  Not Native  Uncommon         NaN                 NaN  

[119248 rows x 13 columns]

In [3]:
#entries where "Record Status" is a Common Name clearly have errors
#entries with "In Review" record status are missing data in some fields,
#may not be as reliable as "Approved" entries
#BUT they make up a substantial portion of the dataset - let's try keeping them in

species_df = species_df[species_df['Record Status'].isin(["Approved", "In Review"])]
species_df

Species ID             Park Name        Category          Order  \
0       ACAD-1000  Acadia National Park          Mammal   Artiodactyla   
1       ACAD-1001  Acadia National Park          Mammal   Artiodactyla   
2       ACAD-1002  Acadia National Park          Mammal      Carnivora   
3       ACAD-1003  Acadia National Park          Mammal      Carnivora   
4       ACAD-1004  Acadia National Park          Mammal      Carnivora   
...           ...                   ...             ...            ...   
119243  ZION-2791    Zion National Park  Vascular Plant      Solanales   
119244  ZION-2792    Zion National Park  Vascular Plant        Vitales   
119245  ZION-2793    Zion National Park  Vascular Plant        Vitales   
119246  ZION-2794    Zion National Park  Vascular Plant  Zygophyllales   
119247  ZION-2795    Zion National Park  Vascular Plant  Zygophyllales   

                Family         Scientific Name  \
0             Cervidae             Alces alces   
1             Cervidae  Odocoileus virginianus   
2              Canidae           Canis latrans   
3              Canidae             Canis lupus   
4              Canidae           Vulpes vulpes   
...                ...                     ...   
119243      Solanaceae       Solanum triflorum   
119244        Vitaceae         Vitis arizonica   
119245        Vitaceae          Vitis vinifera   
119246  Zygophyllaceae       Larrea tridentata   
119247  Zygophyllaceae     Tribulus terrestris   

                                             Common Names Record Status  \
0                                                   Moose      Approved   
1       Northern White-Tailed Deer, Virginia Deer, Whi...      Approved   
2                                  Coyote, Eastern Coyote      Approved   
3             Eastern Timber Wolf, Gray Wolf, Timber Wolf      Approved   
4       Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...      Approved   
...                                                   ...           ...   
119243                                Cut-Leaf Nightshade      Approved   
119244                                       Canyon Grape      Approved   
119245                                         Wine Grape      Approved   
119246                                      Creosote Bush      Approved   
119247                                      Puncture Vine      Approved   

           Occurrence  Nativeness Abundance Seasonality Conservation Status  
0             Present      Native      Rare    Resident                 NaN  
1             Present      Native  Abundant         NaN                 NaN  
2             Present  Not Native    Common         NaN  Species of Concern  
3       Not Confirmed      Native       NaN         NaN          Endangered  
4             Present     Unknown    Common     Breeder                 NaN  
...               ...         ...       ...         ...                 ...  
119243        Present      Native  Uncommon         NaN                 NaN  
119244        Present      Native  Uncommon         NaN                 NaN  
119245        Present  Not Native  Uncommon         NaN                 NaN  
119246        Present      Native      Rare         NaN                 NaN  
119247        Present  Not Native  Uncommon         NaN                 NaN  

[119191 rows x 13 columns]

In [4]:
#what are the possible labels for the "Abundance" column?
species_df["Abundance"].value_counts()

Unknown       28118
Uncommon      18779
Common        12913
Rare          10768
Occasional     3312
Abundant       2365
Name: Abundance, dtype: int64

In [5]:
#this data is primarily text
#so we will need to do some label encoding before we can create and train data models

#label encode Park Name
from sklearn.preprocessing import LabelEncoder
class_labels = LabelEncoder()

def LabelEncode (df):
    for col in df:
        df[col] = class_labels.fit_transform(df[col].values)

In [6]:
#label encode species dataframe
LabelEncode(species_df)
species_df

<ipython-input-5-80ab2d7ab1d4>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = class_labels.fit_transform(df[col].values)


Species ID  Park Name  Category  Order  Family  Scientific Name  \
0                0          0         8     41     417             1130   
1                1          0         8     41     417            29730   
2                2          0         8     87     352             7202   
3                3          0         8     87     352             7204   
4                4          0         8     87     352            45436   
...            ...        ...       ...    ...     ...              ...   
119243      119186         55        13    474    2012            40595   
119244      119187         55        13    546    2301            45396   
119245      119188         55        13    546    2301            45411   
119246      119189         55        13    553    2331            23324   
119247      119190         55        13    553    2331            43839   

        Common Names  Record Status  Occurrence  Nativeness  Abundance  \
0              20527              0           4           0          3   
1              22166              0           4           0          0   
2               8773              0           4           1          1   
3              11183              0           0           0          6   
4               3332              0           4           2          1   
...              ...            ...         ...         ...        ...   
119243          9304              0           4           0          4   
119244          6190              0           4           0          4   
119245         34671              0           4           1          4   
119246          8952              0           4           0          3   
119247         24780              0           4           1          4   

        Seasonality  Conservation Status  
0                12                    8  
1                20                    8  
2                20                    5  
3                20                    0  
4                 0                    8  
...             ...                  ...  
119243           20                    8  
119244           20                    8  
119245           20                    8  
119246           20                    8  
119247           20                    8  

[119191 rows x 13 columns]

In [7]:
#scientific names should be unique - each entry already has unique identifier in "Species ID"
#common names are useful for humans but may just muddy the waters for machine learning/modeling

species_df_filtered = species_df.drop(labels = ["Scientific Name", "Common Names"], axis = 1)

species_df_filtered.head()

Species ID  Park Name  Category  Order  Family  Record Status  Occurrence  \
0           0          0         8     41     417              0           4   
1           1          0         8     41     417              0           4   
2           2          0         8     87     352              0           4   
3           3          0         8     87     352              0           0   
4           4          0         8     87     352              0           4   

   Nativeness  Abundance  Seasonality  Conservation Status  
0           0          3           12                    8  
1           0          0           20                    8  
2           1          1           20                    5  
3           0          6           20                    0  
4           2          1            0                    8

In [8]:
#check data types - see if our label encoding behaved
species_df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 119191 entries, 0 to 119247
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype
---  ------               --------------   -----
 0   Species ID           119191 non-null  int64
 1   Park Name            119191 non-null  int64
 2   Category             119191 non-null  int64
 3   Order                119191 non-null  int64
 4   Family               119191 non-null  int64
 5   Record Status        119191 non-null  int64
 6   Occurrence           119191 non-null  int64
 7   Nativeness           119191 non-null  int64
 8   Abundance            119191 non-null  int64
 9   Seasonality          119191 non-null  int64
 10  Conservation Status  119191 non-null  int64
dtypes: int64(11)
memory usage: 10.9 MB


In [9]:
#modeling to predict "Abundance"
#import train_test_split
from sklearn.model_selection import train_test_split

X = species_df_filtered.drop('Abundance', axis=1)
y = species_df_filtered['Abundance']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [10]:
#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [11]:
#set up Logistic Regression model
from sklearn.linear_model import LogisticRegression

regression_ab = LogisticRegression()
regression_ab.fit(X_train, y_train)

/Applications/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [12]:
#check performance of regression model
regression_ab.fit(X_train, y_train)
y_pred = regression_ab.predict(X_test)
regression_ab.score(X_test, y_test)

/Applications/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.5424724191450984

In [13]:
#import MSE - check RMSE
from sklearn.metrics import mean_squared_error as MSE

rmse_log_ab = MSE(y_test, y_pred) ** (1/2)
print(rmse_log_ab)

1.6909483075926537


In [14]:
#not so hot scores
#maybe dropping zero values in "Abundance" column will help

species_df_filtab = species_df_filtered[species_df_filtered['Abundance'] > 0]

species_df_filtab

Species ID  Park Name  Category  Order  Family  Record Status  \
0                0          0         8     41     417              0   
2                2          0         8     87     352              0   
3                3          0         8     87     352              0   
4                4          0         8     87     352              0   
5                5          0         8     87     863              0   
...            ...        ...       ...    ...     ...            ...   
119243      119186         55        13    474    2012              0   
119244      119187         55        13    546    2301              0   
119245      119188         55        13    546    2301              0   
119246      119189         55        13    553    2331              0   
119247      119190         55        13    553    2331              0   

        Occurrence  Nativeness  Abundance  Seasonality  Conservation Status  
0                4           0          3           12                    8  
2                4           1          1           20                    5  
3                0           0          6           20                    0  
4                4           2          1            0                    8  
5                0           0          6           20                    8  
...            ...         ...        ...          ...                  ...  
119243           4           0          4           20                    8  
119244           4           0          4           20                    8  
119245           4           1          4           20                    8  
119246           4           0          3           20                    8  
119247           4           1          4           20                    8  

[116826 rows x 11 columns]

In [15]:
#modeling to predict "Abundance"
#import train_test_split
from sklearn.model_selection import train_test_split

X = species_df_filtab.drop('Abundance', axis=1)
y = species_df_filtab['Abundance']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [16]:
#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [17]:
#set up Logistic Regression model
from sklearn.linear_model import LogisticRegression

regression_ab = LogisticRegression()
regression_ab.fit(X_train, y_train)

/Applications/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [18]:
#check performance of regression model
regression_ab.fit(X_train, y_train)
y_pred = regression_ab.predict(X_test)
regression_ab.score(X_test, y_test)

/Applications/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.5612000342377814

In [19]:
#import MSE - check RMSE
from sklearn.metrics import mean_squared_error as MSE

rmse_log_ab = MSE(y_test, y_pred) ** (1/2)
print(rmse_log_ab)

1.5810711602791006


In [20]:
#slightly better scores; let's try other modeling methods

#Decision Tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [21]:
X = species_df_filtab.drop('Abundance', axis=1)
y = species_df_filtab['Abundance']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [22]:
dt_ab = DecisionTreeClassifier(max_depth=5, random_state=42)
dt_ab.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=5, random_state=42)

In [23]:
y_pred = dt_ab.predict(X_test)

dt_ab_score = accuracy_score(y_test, y_pred)
print(dt_ab_score)

0.607164255756227


In [24]:
#check rmse
rmse_dt_ab = MSE(y_test, y_pred) ** (1/2)
print(rmse_dt_ab)

1.5156181405707203


In [33]:
#Random Forest
from sklearn.ensemble import RandomForestRegressor

#set seed
SEED = 42

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = SEED)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [26]:
rf_ab = RandomForestRegressor(n_estimators = 400, min_samples_leaf = 0.1, random_state = SEED)
rf_ab.fit(X_train, y_train)

RandomForestRegressor(min_samples_leaf=0.1, n_estimators=400, random_state=42)

In [27]:
y_pred = rf_ab.predict(X_test)

rmse_rf_ab = MSE(y_test, y_pred) ** (1/2)
print(rmse_rf_ab)

1.3738678074942903


In [34]:
#neural network imports
import tensorflow
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 42)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

number_features = X_train.shape[1]

In [35]:
#defining the model
model = Sequential()
model.add(Dense(100, activation='relu', kernel_initializer='he_normal', input_shape=(number_features,)))
model.add(Dense(100, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(7, activation='softmax'))

In [36]:
#compile model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [37]:
#run Abundance neural network model
model.fit(X_train, y_train, epochs=250, batch_size=100, verbose=2)

Epoch 1/250
935/935 - 1s - loss: 0.9133 - accuracy: 0.6206
Epoch 2/250
935/935 - 1s - loss: 0.8259 - accuracy: 0.6586
Epoch 3/250
935/935 - 1s - loss: 0.7991 - accuracy: 0.6726
Epoch 4/250
935/935 - 1s - loss: 0.7819 - accuracy: 0.6807
Epoch 5/250
935/935 - 1s - loss: 0.7682 - accuracy: 0.6851
Epoch 6/250
935/935 - 1s - loss: 0.7554 - accuracy: 0.6905
Epoch 7/250
935/935 - 3s - loss: 0.7455 - accuracy: 0.6957
Epoch 8/250
935/935 - 2s - loss: 0.7374 - accuracy: 0.6982
Epoch 9/250
935/935 - 4s - loss: 0.7312 - accuracy: 0.7023
Epoch 10/250
935/935 - 1s - loss: 0.7255 - accuracy: 0.7033
Epoch 11/250
935/935 - 2s - loss: 0.7214 - accuracy: 0.7047
Epoch 12/250
935/935 - 1s - loss: 0.7179 - accuracy: 0.7060
Epoch 13/250
935/935 - 1s - loss: 0.7138 - accuracy: 0.7075
Epoch 14/250
935/935 - 2s - loss: 0.7106 - accuracy: 0.7071
Epoch 15/250
935/935 - 2s - loss: 0.7086 - accuracy: 0.7083
Epoch 16/250
935/935 - 2s - loss: 0.7053 - accuracy: 0.7077
Epoch 17/250
935/935 - 1s - loss: 0.7028 - accura

935/935 - 2s - loss: 0.5576 - accuracy: 0.7676
Epoch 138/250
935/935 - 2s - loss: 0.5575 - accuracy: 0.7662
Epoch 139/250
935/935 - 2s - loss: 0.5585 - accuracy: 0.7659
Epoch 140/250
935/935 - 1s - loss: 0.5577 - accuracy: 0.7682
Epoch 141/250
935/935 - 2s - loss: 0.5573 - accuracy: 0.7676
Epoch 142/250
935/935 - 2s - loss: 0.5560 - accuracy: 0.7670
Epoch 143/250
935/935 - 2s - loss: 0.5572 - accuracy: 0.7668
Epoch 144/250
935/935 - 3s - loss: 0.5570 - accuracy: 0.7671
Epoch 145/250
935/935 - 2s - loss: 0.5564 - accuracy: 0.7676
Epoch 146/250
935/935 - 4s - loss: 0.5557 - accuracy: 0.7684
Epoch 147/250
935/935 - 2s - loss: 0.5550 - accuracy: 0.7681
Epoch 148/250
935/935 - 3s - loss: 0.5549 - accuracy: 0.7680
Epoch 149/250
935/935 - 2s - loss: 0.5535 - accuracy: 0.7675
Epoch 150/250
935/935 - 1s - loss: 0.5539 - accuracy: 0.7691
Epoch 151/250
935/935 - 1s - loss: 0.5534 - accuracy: 0.7688
Epoch 152/250
935/935 - 2s - loss: 0.5533 - accuracy: 0.7686
Epoch 153/250
935/935 - 2s - loss: 0.5

Of the models, the neural network performed best with about 77% accuracy.  Overall, Abundance seems to be a difficult variable to predict - there may not be a strong correlation between it and the other variables.